flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -83,10 +85,22 @@ public void open(
             final int taskIdx = runtimeContext.getIndexOfThisSubtask();
             final long congruence = start + taskIdx;
 
-            long totalNoOfElements = Math.abs(end - start + 1);
-            final int baseSize = safeDivide(totalNoOfElements, stepSize);
-            final int toCollect =
-                    (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+
+            Preconditions.checkArgument(
+                    start < end, "The start value cannot be greater than the end value.");
+
+            // After preventing setting to Long.MAX_VALUE, the length of
+            // Long type will be exceeded after +1
+            final BigInteger totalNoOfElements =
+                    BigInteger.valueOf(end)
+                            .subtract(BigInteger.valueOf(start))
+                            .add(BigInteger.valueOf(1));
+
+            final BigInteger baseSize = totalNoOfElements.divide(BigInteger.valueOf(stepSize));
+            final long toCollect =
+                    totalNoOfElements.remainder(BigInteger.valueOf(stepSize)).longValue() > taskIdx
+                            ? baseSize.add(BigInteger.valueOf(1)).longValue()
```

Comment:
IIUC, `baseSize` can be bigger than `Long.MAX_VALUE` which will make this line prone to errors. Wouldn't it be easier to just limit the upper threshold to `Integer.MAX_VALUE-1`? :thinking:

Reply:
Do you mean to limit the maximum value of the start of the sequence to Integer.MAX_VALUE-1?

Reply:
yes, that's what I was referring to.